osu.Game/Overlays/Chat/ChannelList/ChannelList.cs

```diff
@@ -187,14 +188,50 @@ public ChannelGroup(LocalisableString label)
                         Margin = new MarginPadding { Left = 18, Bottom = 5 },
                         Font = OsuFont.Torus.With(size: 12, weight: FontWeight.SemiBold),
                     },
-                    ItemFlow = new FillFlowContainer<ChannelListItem>
+                    ItemFlow = new ChannelListItemFlow(sortByRecent)
                     {
                         Direction = FillDirection.Vertical,
                         RelativeSizeAxes = Axes.X,
                         AutoSizeAxes = Axes.Y,
                     },
                 };
             }
+
+            public partial class ChannelListItemFlow : FillFlowContainer<ChannelListItem>
+            {
+                private readonly bool sortByRecent;
+
+                public ChannelListItemFlow(bool sortByRecent)
+                {
+                    this.sortByRecent = sortByRecent;
+                }
+
+                public void Reflow() => InvalidateLayout();
+
+                public override IEnumerable<Drawable> FlowingChildren => sortByRecent
+                    ? base.FlowingChildren.OfType<ChannelListItem>().OrderByDescending(i => i.Channel.LastMessageId ?? long.MinValue)
+                    : base.FlowingChildren.OfType<ChannelListItem>().OrderBy(i => i.Channel.Name);
+            }
+
+            public void AddChannel(ChannelListItem item)
+            {
+                ItemFlow.Add(item);
+
+                item.Channel.NewMessagesArrived += newMessagesArrived;
+                item.Channel.PendingMessageResolved += pendingMessageResolved;
+
+                ItemFlow.Reflow();
+            }
+
+            public void RemoveChannel(ChannelListItem item)
+            {
+                item.Channel.NewMessagesArrived -= newMessagesArrived;
+                item.Channel.PendingMessageResolved -= pendingMessageResolved;
+                ItemFlow.Remove(item, true);
+            }
```

Comment:
Two things here:

- I'd consider not subscribing to the events here if there is no reason to (because the channel group that they're in does not sort by recent) - just to reduce amount of work done
- Technically this can leak memory. `Channel` is a non-drawable class, so having these events hooked up like so without anything in `Dispose()` means that `Channel`s could hold references to disposed `ChannelList`s. In real usage it's probably not an issue because the overlay is basically a singleton, but maybe worth protecting against nonetheless?

Reply:
> Technically this can leak memory

Considering only one `ChannelList` is created ever this isn't really a concern, but sure I'll fix it.

Reply:
Both changes applied in 4fcc762.